Add missing methods and superclass to `FieldFile`

mthuurne opened the pull request:
This adds the pickle methods `__getstate__()` and `__setstate__()` to `django.db.models.fields.files.FieldFile`.

mthuurne committed:
Add pickle methods to `FieldFile`

This adds the methods `__getstate__()` and `__setstate__()`.

sobolevn commented on django-stubs/db/models/fields/files.pyi:
```diff
@@ -31,6 +31,8 @@ class FieldFile(File):
     @property
     def closed(self) -> bool: ...
+    def __getstate__(self) -> dict[str, Any]: ...
+    def __setstate__(self, state: dict[str, Any]) -> None: ...
```
You can also add `__eq__` and `__hash__` as methods from https://github.com/django/django/blob/617bcf611f3daa796e4054ba041089ece30a32fc/django/db/models/fields/files.py#L25-L33

and you can also add `AltersData` base class: https://github.com/django/django/blob/617bcf611f3daa796e4054ba041089ece30a32fc/django/db/models/fields/files.py#L17

While we are at it.

mthuurne replied:
Done.

Please check whether you agree with how I annotated `__eq__()`:
- I assumed `hasattr(other, "name")` was a lazy way to check for `FieldFile`, but if it is intended to accept other classes with a `name` attribute, we should define a `Protocol` instead.
- It accepts `str` for backwards compatibility, but it doesn't state that compatibility is deprecated, so I added it as part of the signature.

sobolevn reviewed (commented):
Thanks!

mthuurne committed:
Add `_eq__()` and `__hash__()` to `FieldFile`

mthuurne committed:
Add `AltersData` superclass to `FieldFile`

mthuurne committed:
Add mypy override for `FieldFile.__eq__()`

The method signature indeed doesn't accept everything `__eq__()`
normally would, but neither does the implementation.

sobolevn commented on django-stubs/db/models/fields/files.pyi:
```diff
@@ -33,6 +34,8 @@ class FieldFile(File):
     def closed(self) -> bool: ...
     def __getstate__(self) -> dict[str, Any]: ...
     def __setstate__(self, state: dict[str, Any]) -> None: ...
+    def __eq__(self, other: FieldFile | str) -> bool: ...
```
It can be `object`, because any value (with correct `__eq__`) can be compared with `self.name`

sobolevn commented on django-stubs/db/models/utils.pyi:
```diff
@@ -7,3 +7,5 @@ def make_model_tuple(model: type[Model] | str | tuple[str, str]) -> tuple[str, s
 def unpickle_named_row(names: Iterable[str], values: Iterable[Any]) -> NamedTuple: ...
 def create_namedtuple_class(*names: str) -> type[NamedTuple]: ...
+
+class AltersData: ...
```
```suggestion
class AltersData:
    def __init_subclass__(cls, **kwargs: Any) -> None: ...
```

sobolevn commented on django-stubs/db/models/fields/files.pyi:
```diff
@@ -31,6 +32,10 @@ class FieldFile(File):
     def closed(self) -> bool: ...
+    def __getstate__(self) -> dict[str, Any]: ...
+    def __setstate__(self, state: dict[str, Any]) -> None: ...
+    def __eq__(self, other: FieldFile | str) -> bool: ...  # type: ignore[override]
```
```suggestion
    def __eq__(self, other: object) -> bool: ...
```

mthuurne committed:
Allow equality comparison to any type

Co-authored-by: Nikita Sobolev <mail@sobolevn.me>

mthuurne committed:
Add `AltersData.__init_subclass__()`

Co-authored-by: Nikita Sobolev <mail@sobolevn.me>

sobolevn reviewed (approved):
Thanks!
